src/ToData.purs

```diff
@@ -1,37 +1,111 @@
 module ToData
-  ( class ToData
+  ( Day(..)
+  , class ConstrIndex
+  , class IsIndexed
+  , class ToData
+  , class ToDataArgs
+  , genericIndexFromConstructor
+  , genericToData
+  , indexFromConstructor
   , toData
+  , toDataArgs
+  , constrToIndex
   ) where
 
 import Prelude
 
+import Contract.Prelude (Tuple(..), genericShow)
 import Data.Array as Array
-import Data.BigInt (BigInt)
+import Data.BigInt (BigInt, fromInt)
 import Data.BigInt as BigInt
 import Data.Either (Either(Left, Right))
 import Data.Foldable (class Foldable)
+import Data.Generic.Rep as G
 import Data.List (List)
 import Data.Map (Map)
 import Data.Map as Map
 import Data.Maybe (Maybe(Just, Nothing))
 import Data.Profunctor.Strong ((***))
 import Data.Ratio (Ratio, denominator, numerator)
+import Data.Symbol (class IsSymbol, reflectSymbol, SProxy(..))
 import Data.Tuple.Nested (type (/\), (/\))
 import Data.UInt (UInt)
 import Helpers (uIntToBigInt)
 import Prim.TypeError (class Fail, Text)
+import Type.Proxy (Proxy(..))
 import Types.ByteArray (ByteArray)
 import Types.PlutusData (PlutusData(Constr, Integer, List, Map, Bytes))
 
+data Day = Mon | Tue | Wed | Thurs | Fri | Sat | Sun
+
+derive instance genericDay :: G.Generic Day _
+
+instance showDay :: Show Day where
+  show = genericShow
+
+instance dayToData :: ToData Day where
+  toData = genericToData
+
 class ToData (a :: Type) where
   toData :: a -> PlutusData
 
+class ConstrIndex (a :: Type) where
+  constrToIndex :: Proxy a -> Map String Int
+
+instance dayConstrIndex :: ConstrIndex Day where
+  constrToIndex _ = Map.fromFoldable [ Tuple "Mon" 0 ]
+
+-- Generic
+class ToDataArgs a where
+  toDataArgs :: a -> Array (PlutusData)
+
+instance toDataArgsNoArguments :: ToDataArgs G.NoArguments where
+  toDataArgs _ = []
+
+instance toDataArgsArgument :: ToData a => ToDataArgs (G.Argument a) where
+  toDataArgs (G.Argument x) = [ toData x ]
+
+instance toDataArgsProduct :: (ToDataArgs a, ToDataArgs b) => ToDataArgs (G.Product a b) where
+  toDataArgs (G.Product x y) = toDataArgs x <> toDataArgs y
```

Comment:
As explained in https://harry.garrood.me/blog/write-your-own-generics/ this is just a neat pattern that flattens a skewed Product of Products